### ISSUE TITLE
Undefined value 05 for Input Source

### ISSUE THREAD

**Issue by denilsonsa** [digitaltrails/vdu_controls#34]
When enabling the "input source" control widget, vdu_controls is complaining about [my AOC displays](https://eu.aoc.com/en/products/monitors/q27p1):

![Error message screenshot](https://user-images.githubusercontent.com/121676/211700494-5258b8f5-7851-4aae-982a-0ed366f17f68.png)

> Display 1 Q27P1B:GNXJ7HA015309 feature 60 '(Input Source)' has an undefined value '05'. Valid values are ['01', '03', '05'].
>
> If you want to extend the set of permitted values, you can edit the metadata for Q27P1B:GNXJ7HA015309 in the settings panel.  For more details see the man page concerning VDU/VDU-model config files.

Then the "Input Source" drop-down shows UNKNOWN-05 as one of the options.

![Screenshot of the drop-down options](https://user-images.githubusercontent.com/121676/211700821-1ea69f75-053d-4d10-8c3e-e5a9e9cb9cf5.png)

These monitors have four input ports: D-SUB (VGA), DVI, HDMI, DP

Using vdu_controls version 1.8.3 on Manjaro Linux. Both monitors are connected through DisplayPort.

---

What's the bug, then?

1. It's weird it claims 05 is invalid while still listing 05 as one of the valid values.
2. It should display 4 inputs instead of only 3. But I'm not sure if you can do anything about it.

Do you need any debugging output from `ddcutil`?

**Comment by denilsonsa** [comment]
After adding some debugging code at [VduControlComboBox.__init__](https://github.com/digitaltrails/vdu_controls/blob/912f5864a7fe24664af4205a7313dc17047a3a46/vdu_controls.py#L2304-L2307), I figured out `self.vcp_capability.values` only has two items (`'01'`→`'VGA-1'`, `'03'`→`'DVI-1'`).

Thus, [the error message](https://github.com/digitaltrails/vdu_controls/blob/912f5864a7fe24664af4205a7313dc17047a3a46/vdu_controls.py#L2336-L2350) is misleading, because it first adds the UNKNOWN value to the list of keys, and later it uses the list of keys to build the error message string.

---

There are two bugs:

1. The error message is misleading. I think you can fix this one. ;)
2. The capability values are only listing 2 out of 4 real values. This might need changes to the ddcutil database or something. Can you help me pointing to the right direction?

**Comment by digitaltrails** [comment]
I see the same problem for my LG monitor.  As described in the documentation, such issues can be fixed in the settings panel for the offending monitor, you can edit its capabilities text, remove the incorrect entries, and then save the new config.    

You can also directly edit the relevant $HOME/.config/vdu_controls/*.conf file if you'd like to use a real text editor.

There is no way around the fact that most monitor vendors do a poor job of implementing DDC.  There are probably too many monitors to expect the author of ddcutil to keep an authoritative list of capabilities (perhaps if ddcutil becomes widely used and a full time job).

**Comment by denilsonsa** [comment]
Thanks! I've edited the capabilities override within vdu_controls and had some success.

Then, inspired by [this comment](https://github.com/ddccontrol/ddccontrol-db/issues/106), I decided to try some values on my AOC Q27P1. It's ugly.

* VGA
    * I don't have any VGA device. Or maybe I do, but it's hard to reach.
    * 0x01 should work fine, but it's untested.
* DVI
    * I don't have any DVI device/cable/adapter.
    * 0x03 should work fine, but it's untested.
* DisplayPort
    * I have to write the value 0x15 to switch to DP.
    * It reports back as 0x05 when reading.
* HDMI
    * I have to write the value 0x17 to switch to HDMI.
    * It reports back as 0x00 when reading.

Switching those inputs takes a few seconds. Meanwhile, vdu_controls tries to reload all the values for all widgets. However, since the display is still blank while switching inputs, trying to read the VCP 10 (Brightness) fails.

![Error message while loading the brightness value](https://user-images.githubusercontent.com/121676/211714472-6061bc27-0ddb-47a8-b2d8-81d7b10a6063.png)

---

New ideas (that could become new issues themselves, what do you think?):

1. When switching inputs, don't reload all the values. Only reload the current one. Or wait a few seconds before reloading. Or don't reload at all.
2. Add support for write-only input source values. Sigh…

**Comment by digitaltrails** [comment]
This might be fixed by increasing the sleep multiplier, at a cost to response time.  

It may be that on any ddcutil error I should back off and wait until the dust settles.  That's what I do at login, because I know things will be slow to sort out during login, I wait until things stabilise before fully initialising.   Perhaps I should do that for any ddcutil error - stay quiet about it, sleep a bit, try again, give up after a two or three failures in a row.

I'll give this some thought.  Plus I'll try and create a test I can experiment with.

**Comment by digitaltrails** [comment]
In commit 700876937225a1add6611c4597ecff2f7fbb6d62 I've  changed the vdu_controls.py to treat all monitor detection situations the same way. The code will loop for a short space of time waiting to see if the number of monitors stabilises.  It only used to do this at session-start, but I see no harm in doing in all the time.  It will cause a slight delay in these situations, but it's a rare situation.  The bonus is that it simplifies the code.

I'm not certain this will clear up all possible issues to do with changing input-sources, unplugging, or switching off.  This is a situation where things get unpredictable due to timings and DDC implementations.  .  In some cases a popup error may be unavoidable, but I would at least like to avoid vdu_controls falling over and exiting.

I'll run this on my machine for a few days to see if there is any fallout.

**Comment by digitaltrails** [comment]
Make that commit b822bc7ecf86865e93b7eaac9e21ee6f418535fc - some code I was trying to help someone with accidentally got left in the previous commit and broke the preset dialog.

**Comment by digitaltrails** [comment]
These kinds of errors happen if the sleep multiplier is set too low.  It was probably a mistake on my part to set the default to 0.5.  Both my own monitors are now set to 0.7 (very old HP ZR24w) and 1.0 (very new LG 4K 27UL60).  I should probably increased the shipped default, after all, that's what ddcutil uses by default.

It was only subsequent to my writing vdu_controls that multi-value setting was added to ddcutil.  I've thought about making use of it, but there are a few issues:

- Do I continue to support those using older versions of ddcutil (presumably those folk on non-rolling distributions)?     
- If the sleep multiplier is set too low, there may still be problems with speed and reliability. I presume ddcutil is still sending each op separately to each monitor, so if sent too fast, any one of them could cause an error?
- What happens if there is an error part way through the sequence, do I replay the entire sequence?

Refactoring the code is also an issue.  I think a low impact change would be to make the ddcutil wrapper code transactional, begin-transaction, set-value, set-value, .. , commit-transaction.

Breaking a lot of things to get back to where I am now, and then find no real improvement in speed would be annoying.   I guess thes best way to start would to be experiment with shell scripts that do the same ops and see if there is a speed and reliability improvement using one approach over the other.

**Comment by denilsonsa** [comment]
I haven't looked at how `ddcui` works. Maybe it calls `ddcutil`, I don't know. But it felt faster and more reliable than `vdu_controls`.

My guess is that `ddcutil` has to detect the devices, then open the i2c device, then send some initial handshake, and only afterwards it can get/set the required values. And after that it might need some closing. My assumption is  that the overhead of launching (and closing) `ddcutil` is too high (causing slowdown) and also leaves the devices in a confused (or blocked) state shortly after the process exits (making the subsequent execution more likely to fail).

And, look, on my experiment (from that animated GIF), I haven't even lowered the default sleep-multiplier, and it was still both fast and reliable (granted, it was not statistically significant, so in fact it "felt" more reliable). And by that I mean the execution didn't crash at all, while `vdu_controls` was throwing errors almost every refresh.

Bonus points to make it even faster: run it in parallel for each display.

Also, if refactoring, it would be better to be non-blocking. Currently the UI gets blocked waiting for `ddcutil`. On the other hand, if it's fast enough, maybe it will never block for long enough.

TL;DR: Although it is quite a lot of work, I'm hoping the end result will be much better.

**Comment by digitaltrails** [comment]
> I haven't looked at how `ddcui` works. Maybe it calls `ddcutil`, I don't know. But it felt faster and more reliable than `vdu_controls`.
> 

It uses a library.  Rockowitz was originally experimenting with PyQt and a python library-wrapper, but apparently the Python/C++ wrapper presented too many issues.

> My guess is that `ddcutil` has to detect the devices, then open the i2c device, then send some initial handshake, and only afterwards it can get/set the required values. And after that it might need some closing. My assumption is that the overhead of launching (and closing) `ddcutil` is too high (causing slowdown) and also leaves the devices in a confused (or blocked) state shortly after the process exits (making the subsequent execution more likely to fail).
> 

My assumption is that the exec of the C++ can't be the primary cost.  As you've written, it's probably the handshaking with the monitors that is causing the delays. 

> And, look, on my experiment (from that animated GIF), I haven't even lowered the default sleep-multiplier, and it was still both fast and reliable (granted, it was not statistically significant, so in fact it "felt" more reliable). And by that I mean the execution didn't crash at all, while `vdu_controls` was throwing errors almost every refresh.

The default sleep multiplier for ddcutil is 1.0.  It was probably a mistake to make it 0.5 in vdu_controls.

> 
> Bonus points to make it even faster: run it in parallel for each display.
> 
> Also, if refactoring, it would be better to be non-blocking. Currently the UI gets blocked waiting for `ddcutil`. On the other hand, if it's fast enough, maybe it will never block for long enough.
> 
> Regarding support for old versions: people on older distros on older versions should use the older `vdu_controls`.

But based on feedback, I suspect the number of users could be counted on one or two hands, so perhaps it's not a big deal.

My primary objective is to alter the brightness of my monitor as the light changes during the day.  That's why there is so much code devoted to presets, schedules, and sun-elevations.   My next project was to solder up a arduino light sensor and side-step all that code by just measuring the ambient light. 

> TL;DR: Although it is quite a lot of work, I'm hoping the end result will be much better.

Rather than make vdu_controls more complex, it may be better to take the ddcui codebase and see if it can be skinned.  But I've had enough C/C++ programming to last a lifetime.  

I did originally have in mind the anyone wanting to do more than brightness/contrast, especially anyone thinking about color management, should probably look at ddcui.   I don't really want to undermine or compete with the efforts of Rockowitz.  I feel vdu_controls should be a light system tray utility.    

I do think the initialisation/refresh should probably use some a single call to ddcutil.  That shouldn't be too hard.

**Comment by digitaltrails** [comment]
I've made an initial stab at getting the refresh to do a single getvcp for each monitor.  It works and the alterations aren't too ugly.   I need to look at doing the same during initialisation.   If this can also be done tidily, then I would hope to have this done by early next week.

Startup may still be a bit slow, because I have to keep probing for monitors until things settle down - typically this is a login time issue when the monitors are too busy to respond properly.  I guess we'll see.

**Comment by digitaltrails** [comment]
I found I could leverage the refresh_data changes mentioned in the previous comment. I've removed the individual getvcp calls made in the initialisation of each control.  Now, when all the controls are layed out, I call just call refresh.   So this change is code complete.  It seems to work, but needs more testing.  I checked in and pushed.   It does subjectively seem faster.  

Time to clock off for the day.

**Comment by denilsonsa** [comment]
> It seems to work, but needs more testing. I checked in and pushed. It does subjectively seem faster.

On a quick test, it seems to work. The longest waiting period is re-detecting the displays. Maybe this can be split into a separate action, so the actual refresh becomes very quick.

**Comment by digitaltrails** [comment]
> > It seems to work, but needs more testing. I checked in and pushed. It does subjectively seem faster.
> 
> On a quick test, it seems to work, and seems faster. The longest waiting period is re-detecting the displays. Maybe this can be split into a separate action, so the actual refresh becomes very quick.

Yes, I noticed that detect seemed to be the slow part.  I might be able to speed up the detect.  It seems to be doing more work around capabilities than I expected.  If the actual connected monitors haven't changed, there is no need to get the capabilities again.

**Comment by digitaltrails** [comment]
> > > It seems to work, but needs more testing. I checked in and pushed. It does subjectively seem faster.
> > 
> > 
> > On a quick test, it seems to work, and seems faster. The longest waiting period is re-detecting the displays. Maybe this can be split into a separate action, so the actual refresh becomes very quick.
> 
> Yes, I noticed that detect seemed to be the slow part. I might be able to speed up the detect. It seems to be doing more work around capabilities than I expected. If the actual connected monitors haven't changed, there is no need to get the capabilities again.

Actually, for me, what was slow was all the loading configs/presets and applying the presets.  Detecting the monitors was comparatively snappy.   

What I have done is remove an the getvcp done before each setvcp.  I was trying to be conservative about writing back to the monitor, by only doing a setvcp if a value actually needs changing.  But that's perhaps a bit too conservative, it's not like there are millions of writes, and it's quite possible some monitor firmware will throw away noop value changes.

**Comment by denilsonsa** [comment]
I had a crazy idea… What if `vdu_controls` used Python's [ctypes](https://docs.python.org/3.11/library/ctypes.html) to interact with `libddcutil.so`? Maybe that would mean initialization once, and less overhead overall (i.e. faster). Still, I have no idea if this is feasible or if this would bring any benefit anyway.

**Comment by digitaltrails** [comment]
A wrapper was how rockowitz first intended to write ddcui in python.  But He found it unworkable:

    I had originally thought that ddcui would be implemented in Python, likely with PyQt.  However, putting a Python wrapper on 
    the C API proved problematic.  I explored SWIG, cffi, cython, and g_object introspection.  In each case, much of the wrapping 
    could be automated, but each had some hand-crafted part that would have to be manually updated with each API change.  I 
    decided it was easier to just implement ddcui in C++.
 
I had considered bolting on a web-service or some other light weight protocol (UDP?) on the front of the C/C++ library.   This would then quite well separate the two.   But I really don't think this is necessary.  I think the speed bottlenecks are more to do with not being able to rely on DDC to always work or to always work promptly.  For example, if a monitor just happens to be busy (maybe it's just been powered on, or maybe the user just logged in), it may not respond correctly.  And this varies from monitor to monitor.   We would just wind up doing sleep(2), repeat, in C instead of python.

The even crazier idea is to ask chatgp for help, I just asked it:

- Can you write a python qt gui front end to ddcutil for controlling VDU brightness and contrast.
- Can you provide an implementation of set_brightness that executes ddcutil in a sub-process.
- Can you do the same for contrast.

The result of the three questions is:

```
import sys, subprocess
from PyQt5.QtCore import Qt
from PyQt5.QtWidgets import QApplication, QWidget, QLabel, QVBoxLayout, QHBoxLayout, QSlider
"""
Can you write a python qt gui front end to ddcutil for controlling VDU brightness and contrast.
Can you provide an implementation of set_brightness that executes ddcutil in a sub-process.
Can you do the same for contrast.
"""

class BrightnessContrastGUI(QWidget):
    def __init__(self):
        super().__init__()
        self.initUI()

    def initUI(self):
        brightness_label = QLabel("Brightness")
        contrast_label = QLabel("Contrast")

        brightness_slider = QSlider(Qt.Horizontal)
        contrast_slider = QSlider(Qt.Horizontal)

        brightness_slider.valueChanged.connect(self.set_brightness)
        contrast_slider.valueChanged.connect(self.set_contrast)

        hbox1 = QHBoxLayout()
        hbox1.addWidget(brightness_label)
        hbox1.addWidget(brightness_slider)

        hbox2 = QHBoxLayout()
        hbox2.addWidget(contrast_label)
        hbox2.addWidget(contrast_slider)

        vbox = QVBoxLayout()
        vbox.addLayout(hbox1)
        vbox.addLayout(hbox2)

        self.setLayout(vbox)

        self.setGeometry(300, 300, 300, 150)
        self.setWindowTitle('Brightness and Contrast')
        self.show()

    def set_brightness(self, value):
        try:
            subprocess.run(["ddcutil", "setvcp", "10", str(value)], check=True)
        except subprocess.CalledProcessError as e:
            print(f"Error setting brightness: {e.output}")

    def set_contrast(self, value):
        try:
            subprocess.run(["ddcutil", "setvcp", "12", str(value)], check=True)
        except subprocess.CalledProcessError as e:
            print(f"Error setting contrast: {e.output}")


if __name__ == '__main__':
    app = QApplication(sys.argv)
    ex = BrightnessContrastGUI()
    sys.exit(app.exec_())
```

I don't give it full marks, I had to add a couple imports.

It does also give some code for _can you write a python ctypes wrapper for libddcutil_, but I'm not in a position to judge that.

I have noticed that when asked about history it sometimes just makes something up that is plausible but totally untrue.

**Comment by digitaltrails** [comment]
I think most of the actionable points in this issue have been addressed.
